contracts/Lottery.sol

```diff
@@ -1,393 +1,648 @@
+    )
+        external
+    {
+        // Ensuring the lottery is within a valid time
+        require(
+            getCurrentTime() >= allLotteries_[_lotteryID].startingTimestamp,
+            "Invalid time for mint:start"
+        );
+        require(
+            getCurrentTime() < allLotteries_[_lotteryID].closingTimestamp,
+            "Invalid time for mint:end"
+        );
+        // Temporary storage for the check of the chosen numbers array
+        uint256 numberCheck = _numberOfTickets.mul(sizeOfLottery_);
+        // Ensuring that there are the right amount of chosen numbers
+        require(
+            _chosenNumbersForEachTicket.length == numberCheck,
+            "Invalid chosen numbers"
+        );
+        // Getting the cost and discount for the token purchase
+        (
+            uint256 totalCost, 
+            uint256 discount, 
+            uint256 costWithDiscount
+        ) = this.costToBuyTicketsWithDiscount(_lotteryID, _numberOfTickets);
+        // Transfers the required cake to this contract
+        cake_.transferFrom(
+            msg.sender, 
+            address(this), 
+            costWithDiscount
+        );
+        // Batch mints the user their tickets
+        uint256[] memory ticketIds = nft_.batchMint(
+            msg.sender,
+            _lotteryID,
+            _numberOfTickets,
+            _chosenNumbersForEachTicket,
+            sizeOfLottery_
+        );
+        // Emitting event with all information
+        emit NewBatchMint(
+            msg.sender,
+            ticketIds,
+            _chosenNumbersForEachTicket,
+            totalCost,
+            discount,
+            costWithDiscount
+        );
     }
 
-    function getTotalRewards(uint256 _issueIndex) public view returns(uint256) {
-        require (_issueIndex <= issueIndex, '_issueIndex <= issueIndex');
 
-        if(!drawed() && _issueIndex == issueIndex) {
-            return totalAmount;
-        }
-        return historyAmount[_issueIndex][0];
+    function claimReward(uint256 _lotteryId, uint256 _tokenId) external {
+        // Checking the lottery is in a valid time for claiming
+        require(
+            allLotteries_[_lotteryId].endTimestamp <= getCurrentTime(),
+            "Wait till end to claim"
+        );
+        // Checks the lottery winning numbers are available 
+        require(
+            allLotteries_[_lotteryId].lotteryStatus == Status.Completed,
+            "Winning Numbers not chosen yet"
+        );
+        require(
+            nft_.getOwnerOfTicket(_tokenId) == msg.sender,
+            "Only the owner can claim"
+        );
+        // Sets the claim of the ticket to true (if claimed, will revert)
+        require(
+            nft_.claimTicket(_tokenId),
+            "Numbers for ticket invalid"
+        );
+        // Getting the number of matching tickets
+        uint8 matchingNumbers = _getNumberOfMatching(
+            nft_.getTicketNumbers(_tokenId),
```

Comment:
Nope. If the ticket was not bought for the specific lottery it will revert above this:
```
require(
            nft_.claimTicket(_tokenId, _lotteryId),
            "Numbers for ticket invalid"
        );
```

The `NFT.claimTicket` checks if the ticket is for the lottery. Will not get to the number of matching line.